# Feature Counts

This analysis uses the *featureCounts* program from the *Subread* software to assign all the sequence reads uniquely aligned by the *STAR* aligner in the <u>Sequence Alignment</u> analysis to their corresponding reference genes for each sequenced sample.

## Inputs

The inputs of this analysis include:

- A set of sequence alignment data files (**.bam*) in the `Aligns` directory.
- The annotation file of UCSC human genome reference library `RefSeq.hg38.gtf` in the `References` directory.

## Procedure

The procedure of this analysis includes the following steps:

1. Count the number of uniquely aligned sequence reads mapped to each reference gene, by running `Programs/Feature-Counts/Count-Conv-RNAseq-Reads.GEO.sh`.

2. Convert the gene read-counts table generated by the *featureCounts* program to the format of the input data file used by the <u>Differential Comparison</u> analysis, by running `Rscript Programs/Feature-Counts/Convert-Conv-RNAseq-Reads.GEO.R`.


## Outputs

The outputs of this analysis include:

- The gene read-counts table `Conv-RNAseq-Read-Counts.tsv` in the `Counts` directory.

