# Convert the read counts of multiple samples generated by featureCounts program
# from Subread software, according to specified experiment design file, based on
# the assumption:
#
# The names of 3'-DGE sequence alignment files have a form of:
#
# RNAseq_[Date].[Well].umi.[sam|bam]
#
# Therefore, the columns of read counts matrix can be match with the rows of
# experiment design table by the following steps:
#
# 1) Extract the Well number from the names of sequence alignment files (included
#    in the header row of the read counts files generated by featureCounts), and
#    assign the Well number to corresponding sample columns of read counts matrix.
#
# 2) Ensure all the Well numbers from the experiment design table are included in
#    those from the read counts matrix.
#
# 3) Extract and sort the columns of the read counts matrix according to the rows
#    of the experiment design table by the Well numbers of samples.

# Load required library
library("tools")

# User home directory
user_home <- Sys.getenv("HOME")

# Control parameters.
# Data text type.
data_text_type <- "txt"
# Data table type.
data_table_type <- "tsv"
# Flag of verbose print.
verbose <- TRUE

# Directory for dataset.
dataset_dir <- file.path(user_home, "LINCSData/Datasets/Difference/LINCS.Dataset/Coen-Paper/DGE-GEO-Depot")
stopifnot(dir.exists(dataset_dir))
# Names of dateset series.
dataset_series_names <- c("Dataset-20150409", "Dataset-20150503", "Dataset-20151120", "Dataset-20161108")
# Main name of experiment design files.
exprt_design_file_main_name <- "DGE-RNAseq-Experiment-Design"
# Main name of read counts files.
read_counts_file_main_name <- "DGE-RNAseq-Read-Counts"
# Name pattern of sequence alignment files.
align_file_name_ptrn <- "\\.(sam|bam)$"

# Convert all series of 3'-DGE datasets
for(dataset_series_name in dataset_series_names)
{
    # Set the directory of current dataset series.
    dataset_series_dir <- file.path(dataset_dir, dataset_series_name)
    if(dir.exists(dataset_series_dir))
    {
        # Set the directories for experiment design and read counts files.
        dataset_series_date <- strsplit(dataset_series_name, split="-")[[1]][2]
        # Directory for experiment design files.
        exprt_design_file_dir <- file.path(dataset_series_dir, "Counts")
        stopifnot(dir.exists(exprt_design_file_dir))
        # Directory for read counts files.
        read_counts_file_dir <- file.path(dataset_series_dir, "Counts")
        stopifnot(dir.exists(read_counts_file_dir))
        # Directory for converted read counts files.
        convert_read_counts_file_dir <- file.path(dataset_series_dir, "Counts")
        stopifnot(dir.exists(convert_read_counts_file_dir))

        # Read read counts table.
        read_counts_file_series_main_name <- paste(read_counts_file_main_name, dataset_series_date, sep="-")
        read_counts_file_name <- paste(read_counts_file_series_main_name, data_text_type, sep=".")
        read_counts_file_path <- file.path(read_counts_file_dir, read_counts_file_name)
        stopifnot(file.exists(read_counts_file_path))
        if(verbose) cat("Reading ", read_counts_file_path, "...\n", sep="")
        read_counts <- read.delim(read_counts_file_path, quote="", comment.char="#", check.names=FALSE, stringsAsFactors=FALSE)
        # Move Geneid column to row names.
        rownames(read_counts) <- read_counts$Geneid
        read_counts$Geneid <- NULL
        # Extract read counts columns whose names contain "bam" keyword.
        read_counts <- read_counts[,grepl(align_file_name_ptrn,colnames(read_counts),ignore.case=TRUE)]
        # Clean up sample names by removing leading directory and extract [Well].
        sample_names <- basename(colnames(read_counts))
        colnames(read_counts) <- sapply(strsplit(sample_names,split="\\."), function(x){return(paste0(x[2],collapse="."))})

        # Name pattern of experiment design files.
        exprt_design_file_name_ptrn <- paste0("^", exprt_design_file_main_name, "-", dataset_series_date, ".*\\.", data_table_type)
        # Find all experiment design files for current dataset.
        exprt_design_file_paths <- list.files(path=exprt_design_file_dir, pattern=exprt_design_file_name_ptrn, full.names=TRUE)
        # Exclude symlink files.
        exprt_design_file_paths <- exprt_design_file_paths[nchar(Sys.readlink(exprt_design_file_paths))==0]
        # Convert the read counts file according to each experiment design file in
        # each dataset.
        for(exprt_design_file_path in exprt_design_file_paths)
        {
            # Read the experiment design of multiple samples.
            if(verbose) cat("Reading ", exprt_design_file_path, "...\n", sep="")
            exprt_design <- read.delim(exprt_design_file_path, quote="", comment.char="#", check.names=FALSE, stringsAsFactors=FALSE)
            exprt_design$ID <- paste(exprt_design$State, exprt_design$Cell, exprt_design$Experiment, exprt_design$Dish, exprt_design$Plate, exprt_design$Well, sep=".")
            stopifnot(!any(duplicated(exprt_design$ID)))

            # Extract and sort sample columns according to experiment design.
            stopifnot(all(exprt_design$Well %in% colnames(read_counts)))
            read_counts_subset <- read_counts[,exprt_design$Well]
            colnames(read_counts_subset) <- exprt_design$ID
            # Sort read counts rows by gene names.
            read_counts_subset <- read_counts_subset[order(rownames(read_counts_subset)),]

            # Write converted read counts of selected samples.
            subset_suffix <- gsub(exprt_design_file_main_name, "", file_path_sans_ext(basename(exprt_design_file_path)))
            convert_read_counts_file_series_main_name <- paste0(read_counts_file_main_name, subset_suffix)
            convert_read_counts_file_name <- paste(convert_read_counts_file_series_main_name, data_table_type, sep=".")
            convert_read_counts_file_path <- file.path(convert_read_counts_file_dir, convert_read_counts_file_name)
            if(verbose) cat("Writing ", convert_read_counts_file_path, "...\n", sep="")
            write.table(read_counts_subset, file=convert_read_counts_file_path, quote=FALSE, na="", sep="\t", row.names=TRUE, col.names=TRUE)
        }
    }
    else
    {
        warning(paste("The dataset directory", dataset_series_dir, "doesn't exist!"))
    }
}
